### PR TITLE
Add Bloat! (also configurable filter lists and lenient protocol matching.)

### PR DESCRIPTION
**Pull request by JoshTheDerf**
Changes:
- Add options page for customizing the filter lists.
- Add "match protocol" option. When true, behaves as before. When false, the protocol is ignored while matching URLs, allowing for redirects such as https://_.fastly.net -> http://_fastly.net and also removing the need for a wildcard in the protocol.
- Update default filter list. ajax.useso.com hosts a Chinese clone of ajax.googleapis.com which should be sufficient.
- Options page and popups page now use [PhotonUI](https://wanadev.github.io/PhotonUI/demo/) because I was too lazy to write all the extra css needed for the widgets.
- Lots more bloat. The urlParse library (A modified copy of Node.js's URL module) could probably be stripped down to just the protocol methods. jQuery could probably be removed as well with little effort.

Be warned though, all of the options.js and popup.js files are entirely undocumented. (Since they're rather long and repetitive, it's just like reading Swing GUIs. :) )

### CODE REVIEW

**Comment by Nateowami** [comment]
Oh, wow, that's a lot of code (6.5x+)! One big commit! I'm guessing some of that was generated though. It's a little late, but I'll take a look at it when I have a moment. Thanks for the contribution! :+1:

**Comment by JoshTheDerf** [comment]
@Nateowami Haha yeah, It's the dependencies. Hence the bloat.

Have fun! I'll be installing this for our friends when they arrive, hopefully.

**Comment by Nateowami** [comment]
This is great! Just a few thoughts and questions before merging:
- Presumably Fastly can't be accessed via https, but http is fine? Can't reproduce this with the home page because they appear to redirect to http by default.
- We should probably add a note about leaving "destination" blank for blocking.
- Great job finding a mirror for ajax.googleapis.com. I'd been wanting that but hadn't found one.
- "Match protocol" needs some explaining. Does it...
  - Only redirect if the "match" URL has the same protocol as the request?
  - Force the redirect to use the same protocol as the request?
  - Or something else?
- Would it be difficult to make the "match protocol" toggle have say "on" or "off," or in some other way give a better visual indicator of the current state?
- What happens when Backloader updates its default filter list, after a user has customized the filters? I'd been hoping we could give filter updates to users, and had deliberately kept the filters separate. Perhaps we could have an "use default filter list" option, and if selected, first match user filters, and then match against the default list.
- "Reset" should probably ask if you're sure. Not a big deal at this time though.
- Nice idea about making the filters reorderable.
- Good idea to update the tag-line to "A Chrome extension to redirect URL requests to other URLs." I need to stop thinking of CDNs exclusively. But I kinda think "URL request" sounds a bit odd. Perhaps "HTTP request"? I don't think that would seem to imply "not https," especially since https is really just HTTP though a secured TCP connection. "To other URLs" is good though.

Yeah, that's long! Thoughts? The ones I'm thinking of the most are the clarification of "match protocol" and the separation of user filters from default filters. Sort of following after Adblock Plus's filter system, with a bit of simplification. :smile_cat:

**Comment by JoshTheDerf** [comment]
:: BEGIN ANSWERS ::
- As far as I can tell (and demonstrated on discuss.atom.io,) fastly.net is blocked (load never starts or completes) over https but works fine over http.
- Good point about that, though I think I stuck it in a tooltip. Perhaps the "Tip:" section could be randomly selected from a list?
- Yay, It's fast too :)
- I'll try to think of a better phrase for Match Protocol. When disabled, it basically matches *:// in the src url, no matter what it is, but redirects to whatever protocol the dest url has. When enabled it enforces the protocol of the src url. Example:

```
// Rule: src: http://googleapis.com, dest: http://useso.com, matchProtocol: false
https://googleapis.com -> http://useso.com
http://googleapis.com -> http://useso.com
pie://googleapis.com -> http://useso.com

// Rule: src: http://googleapis.com, dest: http://useso.com, matchProtocol: true
https://googleapis.com IGNORED
http://googleapis.com -> http://useso.com
```
- Match Protocol can be a [switch (as it is now,) a checkbox, or a toggle button.](http://wanadev.github.io/PhotonUI/demo/CheckBox.html) Which do you prefer?
- I'm still thinking through a way to handle separate filter lists, that is definitely on my TODO list.
- Right, I'll add that to reset. Remind me if I forget.
- Re-orderable filters was something my last attempt sorely needed. Filter lists ought to be ordered by specificity, descending.
- "HTTP requests" would probably work best, if not, then stick with URLs. Oy, gotta love terminology.

:: END ANSWERS ::

**Comment by Nateowami** [comment]
- Regarding the tips: I think there are just a few now, and it might be confusing if the instructions aren't always shown. Perhaps we could just place notes directly under the text box they apply to (e.g. `Use * for wildcards`).
- For protocol matching: Perhaps a checkbox after the text box that says something like "Only match specified protocol."
- Boolean values: I guess I'd go with the checkbox, just because it seems to convey the meaning the best. Not a big deal either way though, but I think the toggle button would be extra confusing in this case.
- Perhaps the default filters could be on another page, but read-only. They might not even have to be added to local storage (not sure about that; been awhile since I've worked with local storage, and I remembering it being a pain).
- Tag line phrasing: Let's go with that then.

One more thing: Photon UI dependencies can't be handled with NPM can they? If they can, it would be nice to keep dependencies out of the repo, but it doesn't really matter that much for a little project like this.

Great work; do you think it's best to merge now, or were you hoping to add another commit first?

**Comment by JoshTheDerf** [comment]
- Alright, we have plenty of space for tips and such in the actual detail view, so I'll move them there.
- Good idea
- Ditto
- This we should probably discuss in detail somewhere else. I have a few ideas but none of which are that easy to implement.
- URL or HTTP request?

PhotonUI is not currently on NPM, so no, it can't really be handled with it yet. :/ (Unless you specify a git repository? I'm not sure how all that works.)

I'm hoping to clean up and implement the simpler bits first before merging. I'll probably get that done sometime this week.

**Comment by Nateowami** [comment]
HTTP request.

Let's continue discussion about separation of filters on #2.

OK, that's fine. Having it in the repo is sub-optimal, but it's not worth too much effort to keep it out.

And we should probably make the next release 0.2.0, following the [semver](http://semver.org/) versioning scheme, since this is new features, not patches.

File: pages/options.js [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+      name: "listButton-"+i
+    });
+    
+    listButton.id = "listButton-"+i,
+    listButton.addClass("listButton");
+    listButton.registerCallback("clicked", "click", function(widget) {
+      buildDetails(filterList[+widget.name.split("-")[1]]);
+      
+      for(var i = 0; i < layout.filterListBox.children.length; i++) {
+        $(".listButton").removeClass("active");
```
Comment (Nateowami) [inline pages/options.js:57]: Why does this need to be executed `children.length` times? Doesn't `$(".listButton")` work globally, i.e., not context-sensitive, so it wouldn't care how many times it gets called? If I'm reading this right the same line gets executed `children.length` times, with no difference between them.

File: pages/options.js [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+      tooltip: "Check the URL's Protocol (http://, https://, etc.) when Testing for This Match.",
+    });
+    
+    var protocolLabel = new p.Label({
+      text: "Match Protocol:",
+      forInput: protocolSwitch,
+      tooltip: "Check the URL's Protocol (http://, https://, etc.) when Testing for This Match.",
+    });
+    
+    layout.applyButton ? layout.applyButton.destroy() : null;
```
Comment (Nateowami) [inline pages/options.js:176]: Isn't this redundant? See [line #137](https://github.com/Nateowami/Backloader/pull/1/files#diff-7f00fc25e8e4066be27c7becea3631e2R137), and again on [line #201](https://github.com/Nateowami/Backloader/pull/1/files#diff-7f00fc25e8e4066be27c7becea3631e2R201). Haven't they already been removed?

File: pages/popup.css [hunk 3]
```diff
@@ -0,0 +1,14 @@
+body {
+	background-color: #FFF;
+	margin: 0px;
+	padding: 0px;
+	width: 130px;
+}
+
+* {
+	-webkit-user-select: none;
```
Comment (Nateowami) [inline pages/popup.css:9]: Woa! Is this the absolutely required for PhotonUI or something? I always wondered how some sites annoyed me that way.

File: scripts/defaultFilterList.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    dest: "http://fonts.useso.com/*",
+    matchProtocol: false,
+  },
+  {
+    src: "https://*fastly.net*",
+    dest: "http://*fastly.net*",
+    matchProtocol: true,
+  },
+  {
+    src: "http://*google.*",
```
Comment (Nateowami) [inline scripts/defaultFilterList.js:23]: Woa! Some of these filters are too strict! We're not allowed to visit http://translate.google.cn/?hl=en-US?

**Comment by Nateowami** [comment]
What about copyright notices for the libraries? That would be PhotonUI, JQuery, array-move?, and FontAwesome. Thoughts about how that should be done?

File: pages/options.js (hunk 2 above)
Comment (JoshTheDerf) [inline pages/options.js:176]: Yep, my bad.

File: pages/popup.css (hunk 3 above)
Comment (JoshTheDerf) [inline pages/popup.css:9]: Haha, No, I just hate accidentally selecting stuff on accident, especially if there was no reason to select stuff in the first place. (I have a strange tendency to click and drag randomly when I'm thinking.)

Plus it feels more desktop-app-ish.

File: scripts/defaultFilterList.js (hunk 4 above)
Comment (JoshTheDerf) [inline scripts/defaultFilterList.js:23]: Oh. That still works. Okay, will change.

**Comment by JoshTheDerf** [comment]
Right, I was rushing, forgot a lot of things, including licensing:
- PhotonUI - BSD-3 clause, should be MIT compatible, so we should be able to just add the notice for it. 
- Array-move is modified off a tutorial on a random site somewhere where I forgot, ages ago. (History digging time.)
- jQuery uses the project's license, or MIT if unspecified, we should be good there.

EDIT: Closing until a more complete version is ready.

File: pages/options.js (hunk 1 above)
Comment (JoshTheDerf) [inline pages/options.js:57]: Yeah, sorry, mistake again. 
Rewrote that a couple of times.

**Comment by Nateowami** [comment]
Not a big deal, just some things I noticed. I haven't been able to find any true bugs. Oh wait. I take that back. If you click "apply," and then mouse off, the brightness resets like it's supposed to, but the bottom border does not. Only matters to nitpickers, but kinda odd.

As for `user-select`... what bugged me was that I couldn't inspect the popup or easily inspect the elements on the options page. But you're not alone with strange habits; I sometimes select text with the keyboard as I read, in order to keep my place. On one site pressing the right arrow skips to the next article... which is super annoying, and has partially helped me break the habit.

Have you found anywhere that talks about how to credit dependencies? I've done it, just by making up my own way, because most resources try to tell you what to do if you want to _modify_ (i.e. fork) the project. I doubt it's much different, but they could at least give some guidelines.

**Comment by JoshTheDerf** [comment]
@Nateowami Oddly enough, that's not a bug. That's how I set up the Less stylesheets. PhotonUI doesn't (yet) have programmatic support for "disabling" widgets, so I added a disabled class to the stylesheet which changes the colors to what you noticed, and disables mouse interaction entirely.

As for the licensing, something like [this](https://github.com/wanadev/PhotonUI/blob/master/AUTHORS) should work, but I have no idea if there's a hard-and-fast way to do it right:

**Comment by Nateowami** [comment]
Oh, I see. Yeah, the disabled class could be improved a bit, but it's no big deal. Thanks for the link regarding licensing; that looks like a pretty good way to do it.

**Comment by Nateowami** [comment]
BTW, I'm happy to help with the changes we've discussed. Not that I'll be able to put much time into it each day (Thursday is my least busy day, and look how much I got done on my main project: a couple lines of code and hit another wall), but if the commits are small enough that we're not trampling on each other's work, we may be able to do this together.

**Comment by JoshTheDerf** [comment]
@Nateowami Sure, I'll also be helping where I can. However, I'm a little tied up at the moment playing dad. :P

**Pull request closed without merging**